docs/development_best_practices.md

```diff
@@ -0,0 +1,130 @@
+
+* When writing a new operator to be used in an application, consider breaking it down into
+	* An abstract operator that encompasses the core functionality but leaves application specific schemas and logic to the implementation
+	* An optional concrete operator also in the library that extends the abstract operator and provides commonly used schema types such as strings, byte[] or POJOs.
+* Follow the conventions of the life cycle methods
+	* Do one time initialization of entities that apply for entire lifetime of the operator in the **setup** method, e.g., factory initializations. Initializations in **setup** are done in the container where the operator is deployed. Initializations in the constructor are discouraged as they lead to extra garbage in memory. This reason is as follows. The operator is instantiated on the client from where the application is being launched, then serialized and made available for launch. So the constructor is first called on the client and if it were to initialize any of the fields, that state would be saved during serialization. In the launch process the operator is deserialized. This would invoke the constructor again, which will initialize the fields but their state will get overwritten by the serialized state and the initial values would become garbage in memory.
+	* Do one time initialization for live entities in **activate** method, e.g., opening connections to a database server or starting a thread for asynchronous operations. The **activate** method is called right before processing starts so it is a better place to do these kind of initializations than at **setup** which can lead to a delay before processing data from the live entity.  
+	* Perform periodic tasks based on processing time in application windows
+	* Perform initializations needed for each application window in **beginWindow**
+	* Perform aggregations needed for each application window  in **endWindow**
+	* Teardown of live entities (inverse of tasks performed during activate) should be in the **deactivate** method
+	* Teardown of lifetime entities (those initialized in setup method) should happen in the **teardown** method
+	* If the operator implementation is not finalized mark it with the **@Evolving** annotation
+* If the operator needs to perform operations based on event time of the individual tuples and not the processing time, extend and use the **WindowedOperator**. Refer to documentation of that operator for details on how to use it.
+* If an operator needs to do some work when it is not receiving any input, it should implement **IdleTimeHandler** interface. This interface contains **handleIdleTime** method which will be called whenever the platform isn’t doing anything else and the operator can do the work in this method. If for any reason the operator is not going to do any work in this method, it should sleep for a small amount of time such as that specified by the **SPIN_MILLIS** attribute otherwise it will cause a busy wait.
```

Comment:
Is it worth adding a caution that work done within this method should not take excessively long, because otherwise the operator may be deemed blocked and killed ?